search-filter.lua

```diff
@@ -0,0 +1,108 @@
+            on_change=self:callback('updateList'), -- Callback function when text in EditField changes
+        },
+        widgets.List{
+            view_id='list',
+            frame={t=3, b=0},
+            choices=self:getFortControlledUnits(), -- Choices in the List view are obtained from getFortControlledUnits function
+            on_select=self:callback('onSelect'), -- Callback function when a unit is selected from the list
+        }
+    }
+end
+
+function SearchEngine:getFortControlledUnits()
+    local fortControlledUnits = {}
+    for _, unit in ipairs(df.global.world.units.active) do
+        local unitName = dfhack.TranslateName(dfhack.units.getVisibleName(unit))
+        if unitName == "" then
+            -- Use the race name if the unit's name field is empty
+            unitName = dfhack.units.getRaceName(unit)
+        end
+        if dfhack.units.isVisible(unit) and dfhack.units.isActive(unit) then
+            table.insert(fortControlledUnits, {text=unitName, search_normalized=dfhack.toSearchNormalized(unitName), id=unit.id})
+        end
+    end
+    table.sort(fortControlledUnits, function(a, b) return a.text < b.text end)
+    return fortControlledUnits
+end
+
+function SearchEngine:updateList()
+    local input = dfhack.toSearchNormalized(self.subviews.edit.text)
+    local fortControlledUnits = self:getFortControlledUnits()
+    local filtered_fortControlledUnits = {}
+
+    for _, unit in ipairs(fortControlledUnits) do
+        if string.find(unit.search_normalized, input) then
+            table.insert(filtered_fortControlledUnits, unit)
+        end
+    end
+
+    self.subviews.list:setChoices(filtered_fortControlledUnits)
+end
+
+function SearchEngine:onSelect(index, unit)
+-- Assign the unit to a specific id
+local unit = df.unit.find(unit.id)
+
+-- Get the position of the unit and center the camera on the unit
+local x, y, z = dfhack.units.getPosition(unit)
+dfhack.gui.revealInDwarfmodeMap(xyz2pos(x, y, z), true)
+
+-- Get the dimensions of the Dwarf Fortress map
+local dims = dfhack.gui.getDwarfmodeViewDims()
+-- Calculate zoom factor based on current viewport zoom level
+local gpsZoom = df.global.gps.viewport_zoom_factor
+-- Set the mouse x and y positions to click on the unit
+df.global.gps.precise_mouse_x = (x - df.global.window_x) * gpsZoom // 4 + gpsZoom // 8
+df.global.gps.precise_mouse_y = (y - df.global.window_y) * gpsZoom // 4 + gpsZoom // 8
+
+-- Enable mouse tracking and set the left mouse button as pressed
+df.global.enabler.tracking_on = 1
+df.global.enabler.mouse_lbut = 1
+df.global.enabler.mouse_lbut_down = 1
+
+-- Simulate a left mouse click at the current mouse position
+gui.simulateInput(dfhack.gui.getDFViewscreen(), '_MOUSE_R')
+
+gui.simulateInput(dfhack.gui.getDFViewscreen(), '_MOUSE_L')
+
+-- Disable mouse tracking and set the left mouse button as not pressed
+df.global.enabler.tracking_on = 0
+df.global.enabler.mouse_lbut = 0
+df.global.enabler.mouse_lbut_down = 0
```

Comment:
```suggestion
```

Reply:
`gui.simulateInput()` also handles mouse state restoration after sending input to the target screen